commons-db/src/main/java/com/palantir/nexus/db/sql/SQLString.java

```diff
@@ -210,75 +219,104 @@ static FinalSQLString getUnregisteredQuery(String sql) {
+     * @param dbType The database type
      */
     private static String makeCommentString(String keyString, DBType dbType) {
         String registrationState;
-        if(keyString != null) {
+        if (keyString != null) {
             registrationState = "SQLString Identifier: " + keyString; //$NON-NLS-1$
         } else {
             registrationState = "UnregisteredSQLString"; //$NON-NLS-1$
         }
         String dbTypeString = ""; //$NON-NLS-1$
-        if(dbType != null)
-         {
+        if (dbType != null) {
             dbTypeString = " dbType: " + dbType; //$NON-NLS-1$
         }
-        String fromDBString = ""; //$NON-NLS-1$
-        return "/* " + registrationState + dbTypeString + fromDBString + " */ "; //$NON-NLS-1$ //$NON-NLS-2$
+        return "/* " + registrationState + dbTypeString + " */ "; //$NON-NLS-1$ //$NON-NLS-2$
+    }
+
+    @VisibleForTesting
+    static String canonicalizeStringAndRemoveWhitespaceEntirely(String sql) {
+        return canonicalizeString(sql, true);
     }
 
     /**
      * Cleans up whitespace, any trailing semicolon, and prefixed comments that a string is
      * unregistered, in order to come up with a canonical representation of this sql string.
-     * @param s
-     * @return
      */
-    public static String canonicalizeString(String s) {
-        String trim = s.trim();
-        if(!trim.isEmpty() && trim.charAt(trim.length() - 1) == ';') {
-            trim = trim.substring(0, trim.length() - 1);
+    public static String canonicalizeString(String sql) {
+        return canonicalizeString(sql, false);
+    }
+
+    private static String canonicalizeString(String original, boolean removeAllWhitespaceEntirely) {
+        StringBuilder cleanedString = new StringBuilder(original);
+        int originalIdx = 0;
+        int cleanedIdx = 0;
+        int firstUnregisteredIdx = cleanedString.indexOf(UNREGISTERED_SQL_COMMENT);
+
+        while (originalIdx < original.length()) {
+            char originalChar = original.charAt(originalIdx);
+            if (originalIdx == firstUnregisteredIdx) {
+                originalIdx += UNREGISTERED_SQL_COMMENT.length();
+                firstUnregisteredIdx = original.indexOf(UNREGISTERED_SQL_COMMENT, originalIdx);
+            } else if ((Character.isWhitespace(originalChar))
+                    && ((cleanedIdx == 0)
+                    || (Character.isWhitespace(cleanedString.charAt(cleanedIdx - 1)))
+                    || removeAllWhitespaceEntirely)) {
+                ++originalIdx;
+            } else {
+                cleanedString.setCharAt(cleanedIdx, originalChar);
+                ++cleanedIdx;
+                ++originalIdx;
+            }
         }
-        String prefix = "/* UnregisteredSQLString */"; //$NON-NLS-1$
-        if (trim.startsWith(prefix)) {
-            trim = trim.substring(prefix.length()).trim();
+
+        while (cleanedIdx > 0 && Character.isWhitespace(cleanedString.charAt(cleanedIdx - 1))) {
```

Comment:
As @SerialVelocity pointed out, this can be an `if` since the checks on lines 283-284 would maintain that the cleaned part of the string can't have consecutive whitespace.